lib/tbot/tbot_test.go

```diff
@@ -0,0 +1,196 @@
+	err := svc.GetAuthServer().RotateCertAuthority(ctx, auth.RotateRequest{
+		// only rotate Host CA as to avoid race condition serverside when
+		// multiple CAs are rotated at once and the database closes off.
+		Type:        types.HostCA,
+		Mode:        "manual",
+		TargetPhase: phase,
+	})
+	if err != nil {
+		log.WithError(err).Infof("Error occurred during triggering rotation: %s", phase)
+	}
+	require.NoError(t, err)
+	log.Infof("Triggered rotation: %s", phase)
+}
+
+func setupServerForCARotationTest(ctx context.Context, log utils.Logger, t *testing.T, wg *sync.WaitGroup) (auth.ClientI, func() *service.TeleportProcess, *config.FileConfig) {
+	fc, fds := testhelpers.DefaultConfig(t)
+
+	// Patch until https://github.com/gravitational/teleport/issues/13443
+	// is resolved.
+	fc.Databases.EnabledFlag = "false"
+
+	cfg := service.MakeDefaultConfig()
+	require.NoError(t, config.ApplyFileConfig(fc, cfg))
+	cfg.FileDescriptors = fds
+	cfg.Log = log
+	cfg.CachePolicy.Enabled = false
+	cfg.Proxy.DisableWebInterface = true
+
+	svcC := make(chan *service.TeleportProcess)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := service.Run(ctx, *cfg, func(cfg *service.Config) (service.Process, error) {
+			svc, err := service.NewTeleport(cfg)
+			if err == nil {
+				svcC <- svc
+			}
+			return svc, err
+		})
+		require.NoError(t, err)
+	}()
+
+	var svc *service.TeleportProcess
+	select {
+	case <-time.After(10 * time.Second):
+		t.Fatal("teleport process did not instantiate in 10 seconds")
+	case svc = <-svcC:
+	}
+
+	// Ensure the service starts correctly the first time before proceeding
+	eventCh := make(chan service.Event, 1)
+	svc.WaitForEvent(svc.ExitContext(), service.TeleportReadyEvent, eventCh)
+	select {
+	case <-eventCh:
+	case <-time.After(30 * time.Second):
+		// in reality, the auth server should start *much* sooner than this.  we use a very large
+		// timeout here because this isn't the kind of problem that this test is meant to catch.
+		t.Fatal("auth server didn't start after 30s")
+	}
+
+	// Tracks the latest instance of the Teleport service through reloads
+	activeSvc := svc
+	activeSvcMu := sync.Mutex{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case svc := <-svcC:
+				activeSvcMu.Lock()
+				activeSvc = svc
+				activeSvcMu.Unlock()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return testhelpers.MakeDefaultAuthClient(t, log, fc), func() *service.TeleportProcess {
+		activeSvcMu.Lock()
+		defer activeSvcMu.Unlock()
+		return activeSvc
+	}, fc
+}
+
+// TestCARotation is a heavy integration test that through a rotation, the bot
+// receives credentials for a new CA.
+func TestBot_Run_CARotation(t *testing.T) {
+	t.Parallel()
+	if testing.Short() {
+		t.Skip("test skipped when -short provided")
+	}
+
+	// wg and context manage the cancellation of long running processes e.g
+	// teleport and tbot in the test.
+	log := libutils.NewLoggerForTests()
+	wg := &sync.WaitGroup{}
```

Comment:
I don't feel super strongly, so feel free to push back, but I generally use a wait group when there's a group (ie more than one) thing I want to wait for.

In this case, a simple channel read would also work, no?

Reply:
There is actually multiple things in the group :') It's passed into `setupServerForCARotationTest` but I haven't really helped my case by obscuring that there. Essentially, it's a group of the bot and the running Teleport process, so that I can ensure both have closed down.